### ISSUE TITLE
onSucess when crossorigin

### ISSUE THREAD

**Issue by AndreaMaestri18** [bhubr/react-simple-oauth2-login#60]
Hello everyone, thanks for the library first of all!

I'm working with a cross-origin request and I need to display the code from the response.

However, when I try it doesn't show up. If I set crossorigin to false I get an error in console about not being able to read the href from because of the cross-origin request. Nevertheless the code in that case is logged.

Has anyone encountered this before?

This is what I'm doing:
`        
<OAuth2Login
          authorizationUrl="authUrl"
          responseType="code"
          clientId={clientId}
          redirectUri={redirectUri}
          onSuccess={(e) => console.log(e)}
          onFailure={(e) => console.log(e)}
          isCrossOrigin={true} //if false, the console.log works
          buttonText={"Login"}
        />
        `

**Comment by bhubr** [comment]
Sorry for the delayed response.

When you say "I don't have any control on the server", you mean the server backing your React app, not the OAuth2 server?

That said, I might have a solution. The `isCrossOrigin` flag is supposed to be used when the `redirectUri` param points to an URL on the _server_.

You might try setting `redirectUri` to your React app's URL. In this case, your React app can extract the code from the URL and send it to your backend app, which then requests a token from the OAuth2 provider, then sends it back to your frontend app. But maybe you tried this already?

**Comment by AndreaMaestri18** [comment]
no i mean the Oauth2 server, I don't have control on that, so I can't set the cors in the header of the response (I'm logging in a different application than my react app). The `redirectUri` that i use is localhost (because I want to be redirected in my app after I log-in)

**Comment by bhubr** [comment]
Ok. Sure, unless it's your own OAuth2 server & infrastructure, you can't control how it behaves regarding CORS.

The thing is, you're supposed to have your own backend app if you want to use the "authorization code" flow. This app can be on your localhost while you're developing. Then there are two cases:

* If your `redirectUri` is set to your server app, you should use `isCrossOrigin`.
* If it's set to your React app, you needn't use `isCrossOrigin`.

**Comment by AndreaMaestri18** [comment]
I have also a backend app yes! But the problem is that if the redirectUri is my react app and I don't use the `isCrossOrigin` I get the error I mentioned before

**Comment by bhubr** [comment]
Is your project open source? If so, I could have a look.

Otherwise I have a few possible hints:
* is CORS configured on *your* backend app? If not, it's one of the possible causes of failure.
* other than configuring CORS on your backend, you might (if you're using Create React App) set the `proxy` setting to your backend's URL, so that when you send an XHR request with your auth code (e.g. POST to `http://localhost:3000/oauth/token`) this request is seamlessly proxied to your server.
* another way would be to do without CORS altogether, by having *both* your backend and frontend under the same domain. It should be possible to do so with a local webserver (e.g. Nginx or Apache running on localhost) with reverse proxies to your backend & frontend apps. If you're familiar with Docker it might even be a bit simpler to achieve.

If you need help with the last option (webserver with or without Docker), let me know what are your OS and server stack (PHP? Node? Python? other? which framework?) and I'll try to come up with the right setup instructions & config files.

**Comment by bhubr** [comment]
Hi @AndreaMaestri18 

Have you progressed on the issue?
Let me know if you need help!

**Comment by Djalalov** [comment]
Hi @bhubr, 

First of all, thanks for contributing for this simple yet very useful component.

**Comment by bhubr** [comment]
Hi @Djalalov 

Thanks! I guess you had a question in mind?

**Comment by Djalalov** [comment]
![Screenshot from 2023-11-29 14-02-45](https://github.com/bhubr/react-simple-oauth2-login/assets/51530302/444c2d7e-7dd6-409e-9938-ee8d52daca9e)

We have the exact scenario described here but we dont understand what it means for the step 2. It says set the auth url in backend and it should return something. Can you please elaborate on this ?

**Comment by bhubr** [comment]
OK I kind of see,

I need to see some digging: the `isCrossOrigin` stuff had been added by a contributor which had this specific use case.

I had started a big overhaul of the docs but never got through it, and I definitely should get back to it.

But first, I'm gonna try and retrieve the code samples I think I have somewhere. I have some time right now so I'm gonna get right to it. Stay tuned, I'll keep you posted!

**Comment by Djalalov** [comment]
![image](https://github.com/bhubr/react-simple-oauth2-login/assets/51530302/ab51d2f1-565c-4465-ac25-b6328df937df)
Here is the screenshot

**Comment by bhubr** [comment]
Might I ask you some clarification please?

When you "authorize" your OAuth app in the popup window, are you **directly** redirected to <http://localhost:8080> with the access token? What's what your screenshot indicates...

What I mean is, don't you first get a "code" that you then "exchange" for an access token by sending a POST request to a specific endpoint on Jira?

**Comment by Djalalov** [comment]
> **[EDIT] SOLVED! See comment below!**
> 
> Might I ask you some clarification please?
> 
> When you "authorize" your OAuth app in the popup window, are you **directly** redirected to http://localhost:8080/callback with the access token? That's what your screenshot seems to indicate...
> 
> What I mean is, don't you first get a "code" that you then "exchange" for an access token by sending a POST request to a specific endpoint on Jira?

Yes, its where things get complicated. What you are saying is the standerd flow and this was our plan at first IDEALLY. But that didnt work since we have CORS issue. We tried move the logic of exchanging the code to token in the backend right after when Jira hits 8080/callback

**Comment by bhubr** [comment]
> > **[EDIT] SOLVED! See comment below!**
> > Might I ask you some clarification please?
> > When you "authorize" your OAuth app in the popup window, are you **directly** redirected to http://localhost:8080/callback with the access token? That's what your screenshot seems to indicate...
> > What I mean is, don't you first get a "code" that you then "exchange" for an access token by sending a POST request to a specific endpoint on Jira?
> 
> Yes, its where things get complicated. What you are saying is the standerd flow and this was our plan at first IDEALLY. But that didnt work since we have CORS issue. We tried move the logic of exchanging the code to token in the backend right after when Jira hits 8080/callback

Interesting. Do you mean that you initially intend to redirect to localhost:3002/callback, then send the code to your backend, have it request the token and send it back to your React app?

I guess you already tried configuring CORS on the backend? I don't know how that would be done with Java (Spring Boot maybe?) but it's also a common issue when dealing with requests from a React app on a Node.js backend.

But then, there are ways to fix it without necessarily doing cross-origin requests:

* One would be to use the `"proxy"` param in `package.json` - **it should work if you're using Create React App**, but I'm not sure if it will with other bundlers/build systems (Vite, etc.)
* Another would be to dockerize the whole thing: backend + React app, though as I'm writing it, I'm having kind of a doubt. I know it works in production, but I'm unsure about it in development. The idea is to put both the backend and the frontend behind an nginx reverse proxy, with all backend traffic going to e.g. port 8080 (it would require to have all backend endpoints below a common prefix, e.g. `/api`) and all frontend traffic (everything that's not `/api`) going to .e.g. 3002. I think I could provide an example if needs be (that would help me make sure that it's actually doable - or not).

**Comment by bhubr** [comment]
@Djalalov Did you eventually succeed?

**Comment by Djalalov** [comment]
> @Djalalov Did you eventually succeed?

Hi @bhubr, yes. 

After several attempts, we have just achieved desired result.

![image](https://github.com/bhubr/react-simple-oauth2-login/assets/51530302/03acb210-71a8-424a-b81a-acc8e1367c5f)

Just for the reference: This snippet is the key. Just because after Jira redirect user to the endpoint 8080/callback with code and state, backend was literally **catching** that response from Jira, exchanging code and state to access_token and not doing anything else. You snippet of code was actually the code we needed to send the data to the front. 

I really appreaciate your help. Thanks

**Comment by bhubr** [comment]
@Djalalov Glad it helped!
